oeps/oep-38-Data-Modeling.rst

```diff
@@ -0,0 +1,442 @@
+
+-  Use a numeric foreign key(s)
+
+   -  Data should be joined on other foreign keys
+
+   -  Do not join on things such as username, email address, or other
+         dimensions of data that may change over time.
+
+-  The minimum number of indexes possible to make the table/queries
+      performant
+
+   -  Indexes cost space and have their own set of performance concerns.
+
+   -  Over-indexing data could actually make the database less
+         performant (slower writes/updates)
+
+   -  You should use as many indexes as you need to make the application
+         performant and no more.
+
+-  History for models involved with enrollments, or anything that may
+      have an effect on financial payments or transactions.
+
+-  History using django-simple-history
+
+   -  Should capture the ID of the user who made the change
+
+-  Where Django simple history is not an option, the following data
+      should be captured:
+
+   -  Fields that were changed
+
+   -  Date & time of the change
+
+   -  ID of the user who initiated the change
+
+-  The correct data type for a column.
+
+   -  Don’t use a long when an Int or bool would do.
+
+   -  Use Integers for your keys
+
+   -  Don’t store an int as varchar
+
+-  Don’t mix types in a column
+
+-  Annotations
+
+   -  All models in the OpenedX ecosystem should be tagged for PII
+      using `code
+      annotations <https://github.com/edx/code-annotations>`__\  following \ `OEP-30 <https://github.com/edx/open-edx-proposals/blob/master/oeps/oep-0030-arch-pii-markup-and-auditing.rst>`__
+
+Best Practices 
+~~~~~~~~~~~~~~~
+
+Deleting data:
+
+-  It is better to have a column to mark the record as inactive than to
+      remove the data from the system using the SQL delete keyword.
+
+   -  Please note that GDPR may require the data be deleted. If a field
+         is determined to contain PII and falls under the realm of GDPR,
+         That data should be deleted from the system.
+
+Don’t trap the data
+
+-  Each piece of information should have its own column. Avoid storing
+      data in blob fields or as JSON in the database.
+
+-  Don’t store encoded (pickle, json, other) objects in the database! If
+      you need to run the python environment to decode the data, the
+      data is worthless.
+
+Store everything
+
+-  Storage is cheap!
+
+-  If you are unsure whether you should store something in the database,
+      add history, or fire an event, the answer is almost always yes
+
+-  Still not sure? The default answer is yes.
+
+Views should access models via methods on those models, instead of
+querying managers directly.
+
+-  For example, prefer creating something like
+      CourseEnrollment.is_enrolled(...) rather than having views check
+      CourseEnrollment.objects.filter(...).exists().
+
+-  This allows us to more easily make internal representation changes
+      like switching to using a “deleted” flag instead of deleting the
```

Comment:
Flag is an ambiguous term.  Can we have a stance on flags being booleans with default values, nullable booleans, or timestamps (EG delete_at. Null for active records, timestamped deletes otherwise)

Reply:
I have updated the OEP to direct the reader to use django_utils [SoftDeletableModel](https://django-model-utils.readthedocs.io/en/latest/models.html#softdeletablemodel)

Reply:
I added an example of adding a default value to a boolean and having sane defaults. That said I think there are cases where having None, True and False for values on a boolean are valid (None: I haven't viewed the thing, True: I agree, False: I disagree). I don't think we should take a hard stance here.